docs/rfc/rfc-1.adoc

```diff
@@ -401,12 +385,99 @@ to submit multiple redemptions. After a redemption, if the wallet has under
+
+```
+// wallet pubkey hash to the current main's UTXO hash computed
+// as keccak256(txHash | txOutputIndex | txOutputValue)
+mapping(bytes20 => bytes32) public mainUtxos;
+
+// wallet pubkey hash to the spent main UTXO hash computed
+// as keccak256(txHash | txOutputIndex)
+mapping(bytes20 => mapping(bytes32 => bool)) spentMainUTXOs
+
+```
+
+Note that this approach is stricter than just validating the public key in the
+UTXO. For example, if we were considering that a collection of deposit UTXO’s
+plus the main UTXO for the wallet unlocked using the wallet public key and
+locked under a single UTXO using the same wallet public key is not a fraud but
+a normal sweep, the Bridge would be susceptible to attacks when the malicious
+wallet steals revealed deposits by sweeping them to another UTXO but not the main
+UTXO known by the Bridge.
+
+
+==== Submitting Fraud Challenge
+
+Anyone should be able to submit a fraud challenge. At a minimum, the function
+should accept the wallet's signature and the sighash uniquely identifying UTXO
+unlocked by the wallet. The function should validate the wallet state to make
+sure it's neither closed nor terminated. It should also validate if the signature
+over the sighash is valid and if it belongs to the wallet.
+
+The function should require the challenger to provide a deposit in ETH that is
+returned to the challenger once the fraud is confirmed. If the wallet defeats the
+challenge, the deposit is sent to the treasury.
+
+==== Defeating Against Fraud Challenge
+
+Anyone can defeat the wallet against the fraud challenge by submitting a transaction
+preimage matching the challenged sighash and proving that transaction was valid to
+the protocol.
+
+The function allowing to defend against the challenge, should first validate the
+submitted preimage against challenged sighash, and try to locate the revealed deposit
+for the input corresponding to the sighash.
+
+If there is a revealed deposit matching the sighash and that deposit was swept by the
+Bridge, everything is fine and the wallet defeated the challenge.
+
+If there is a revealed deposit matching the sighash and it was not yet swept by the
+Bridge, the defeat function should revert. The wallet should first submit the sweep
+proof for that deposit and then call the defeat function again.
 
-When a redemption is requested, the redeeming bitcoin public key and amount are
-known on the ethereum chain. Any bitcoin transactions with the main wallet
-wallet UTXO as the input must have outputs that match those known redemption
-requests, otherwise the transaction was fraudulent.
+If the revealed deposit does not exist, the defeat function should check if the sighash
+belongs to one of the wallet's spent UTXOs from `spentMainUTXOs` mapping. If it does,
+everything is fine and the wallet defeated the challenge. If it does not, the defeat
+function reverts. The wallet should submit the redemption proof and then, call the
+defeat function again.
 
```

Comment:
Also, I think it makes sense to not allow for starting closure (emptying) of a wallet which is currently being challenged. If you agree, could we add some info about that in the RFC?

Reply:
> Could we add here information what should happen if wallet does not successfully defeat the challenge before fraud_challenge_defeat_timeout?

It is now explained earlier according to [this comment](https://github.com/keep-network/tbtc-v2/pull/135#discussion_r811893133). Please see  913ffecab643e672ce664c8c7dd8071e33a9e1ca

Reply:
> Also, I think it makes sense to not allow for starting closure (emptying) of a wallet which is currently being challenged.

I think it should be allowed. Otherwise, a malicious wallet can intentionally fail a heartbeat to commit fraud later. Tokens can be slashed from a malicious wallet no matter if there is BTC on that wallet or not.

Reply:
I think your counterexample applies to a different scenario that I'm talking about - it would be a counterexample if I wanted to not allow for submitting the fraud challenge when the wallet is closing. What I meant was the other way round - to not allow for closing of a wallet when there is already a pending fraud challenge for that wallet.

I just want to make sure that the wallet members can't close the wallet and unstake before they get slashed when `fraud_challenge_defeat_timeout` passes.

After what time since closing the wallet will wallet members be able to unstake? Does it work like that?:

Wallet gets emptied and closed.
Wallet members wishing to unstake would first need to decrease the authorization to 0 for the tBTC application, the application would need to approve that authorization decrease (*) and only then would the staker be able to unstake.
And we could implement the application in that way that it would not process the (*) step if wallet has pending fraud challenge?

Reply:
Closing the wallet does not allow its members to unstake immediately. We will have a fixed undelegation period (working on the staking contract integration right now), independent from the fact whether the operator is a member of some wallet or not.

Reply:
OK, I see, I'm fine with resolving this conversation for now, but I will want to look at the undelegation period logic once it's ready, please add me as a reviewer once you have a PR for that.